### PR TITLE
Speedup Process.children() 

### PR DESCRIPTION
**Pull request by giampaolo**
See #1183.

### CODE REVIEW

**Comment by giampaolo** [comment]
```
BEFORE PATCH

~/svn/psutil {master}$ python -m timeit -s "import psutil; p = psutil.Process()" "p.children()"
100 loops, best of 3: 12.2 msec per loop

AFTER PATCH

~/svn/psutil {master}$ python -m timeit -s "import psutil; p = psutil.Process()" "p.children()"
100 loops, best of 3: 5.49 msec per loop
```

**Comment by pitrou** [comment]
Here are the numbers I get:
* git master:
```
$ python -m timeit -s "import psutil; p = psutil.Process()" "list(p.children())"
10 loops, best of 3: 35.3 msec per loop
```
* this PR:
```
$ python -m timeit -s "import psutil; p = psutil.Process()" "list(p.children())"
100 loops, best of 3: 15.5 msec per loop
```
* PR #1184:
```
$ python -m timeit -s "import psutil; p = psutil.Process()" "list(p.children())"
100 loops, best of 3: 13.6 msec per loop
```

So my approach is slightly faster, at the expense of being Linux-only. Of course, one could argue that WIndows and Linux are the dominant platforms ;-)

**Comment by giampaolo** [comment]
> Of course, one could argue that WIndows and Linux are the dominant platforms ;-)

Makes sense =)
Here's a variation of your patch: https://github.com/giampaolo/psutil/pull/1185/commits/78f8cbcf0a93b2b075803ad9578bc317ce599fd7.

File: psutil/_pslinux.py [hunk 1]
```diff
@@ -1356,6 +1356,27 @@ def pid_exists(pid):
...
+    one shot. Used to speed up Process.children().
+    """
+    ret = {}
+    procfs_path = get_procfs_path()
+    for pid in pids():
+        try:
+            with open_binary("%s/%s/stat" % (procfs_path, pid)) as f:
+                data = f.read()
+        except EnvironmentError as err:
+            if err.errno not in (errno.ENOENT, errno.EPERM, errno.EACCES):
```
Comment (pitrou) [inline psutil/_pslinux.py]: Apparently also `ESRCH` (according to `wrap_exceptions`)?

**Comment by giampaolo** [comment]
Merged. Thanks for reporting this. I had no idea such a speedup could be possible.

**Pull request merged**